### PR TITLE
[SPARK-45174][CORE] Support `spark.deploy.maxDrivers`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to support a new configuration, `spark.deploy.maxDrivers`, like the existing `spark.mesos.maxDrivers`.

### Why are the changes needed?

To limit the maximum number of running drivers.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs with the newly added test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this PR when you have some time, @viirya ?

**Review by yaooqinn — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you so much, @yaooqinn !

**Comment by dongjoon-hyun** [comment]
`core` module test passed and the other Python Connector failures are known ones. Let me merge this.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 1]
```diff
@@ -844,8 +845,8 @@ private[deploy] class Master(
       // We assign workers to each waiting driver in a round-robin fashion. For each driver, we
       // start from the last worker that was assigned a driver, and continue onwards until we have
       // explored all alive workers.
-      var launched = false
-      var isClusterIdle = true
+      var launched = (drivers.size - waitingDrivers.size) >= maxDrivers
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala:848]: Hmm, isn't this making submitting driver silently ignored? For example, there is a warning log for the case the driver needs more resources than any worker. Should we also have a warning log for the case that drivers reaches the maximum drivers?

**Review by viirya — COMMENTED** [review]
Looks good with a minor comment on adding possible logging.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/Master.scala:848]: No~ `schedule` is re-trigger in many cases and especially on the removal of running drivers. So, if the number of running drivers dropped down, new drivers will be allocated to the slots, @viirya .

**Comment by dongjoon-hyun** [comment]
For the log message, the log message could be very verbose because `schedule` method is invoked at every submission.

For example, if we submit 500 jobs with the max limi 10, 490 line of the same logs will be printed. In any way, it's a kind of queuing.

Thank you for review, @viirya .

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/deploy/master/Master.scala:848]: Got it. I just worry about the case that submitted drivers are silently waiting there without any feedback to users.

**Comment by viirya** [comment]
Yea, the verbose logging messages could be an issue.

**Pull request closed without merging**